modules/connectionquality/connectionquality.js

```diff
@@ -68,23 +134,41 @@ export default {
      * @param id the id associated with the statistics
      * @param data the statistics
      */
-    updateRemoteStats: function (id, data) {
-        if (!data || !("packetLoss" in data) || !("total" in data.packetLoss)) {
-            eventEmitter.emit(CQEvents.REMOTESTATS_UPDATED, id, null, null);
-            return;
-        }
-        // Use only the fields we need
-        data = {bitrate: data.bitrate, packetLoss: data.packetLoss};
+    updateRemoteStats:
+        function (id, data, remoteVideoType, isRemoteVideoMuted) {
+            if (!data ||
+                !("packetLoss" in data) ||
+                !("total" in data.packetLoss)) {
+                eventEmitter.emit(CQEvents.REMOTESTATS_UPDATED, id, null, null);
+                return;
+            }
+
+            let inputResolution = data.resolution;
+            // Use only the fields we need
+            data = {bitrate: data.bitrate, packetLoss: data.packetLoss};
 
-        remoteStats[id] = data;
+            remoteStats[id] = data;
 
-        var newVal = 100 - data.packetLoss.total;
-        var oldVal = remoteConnectionQuality[id];
-        remoteConnectionQuality[id] = calculateQuality(newVal, oldVal || 100);
+            if (disableQualityBasedOnBandwidth
+                || isRemoteVideoMuted
+                || remoteVideoType === 'desktop'
+                || !inputResolution) {
+                var newVal = 100 - data.packetLoss.total;
+                var oldVal = remoteConnectionQuality[id];
+                remoteConnectionQuality[id]
+                    = calculateQuality(newVal, oldVal || 100);
+            } else {
+                let val = calculateQualityUsingUpload(
+                    inputResolution.inputHeight,
+                    data.bitrate.upload);
+                if (val) {
+                    remoteConnectionQuality[id] = val;
+                }
+            }
```

Comment:
I think this `if` is exactly the same as the one for the local values. Can we somehow include that logic in function called from both places?